cni/pkg/ambient/net_linux.go

```diff
@@ -1189,9 +1190,6 @@ func (s *Server) CreateRulesWithinNodeProxyNS(proxyNsVethIdx int, ztunnelIP, ztu
 func (s *Server) cleanupNode() {
 	log.Infof("Node-level network rule cleanup started")
 	if s.redirectMode == EbpfMode {
-		if err := s.delZtunnelEbpfOnNode(); err != nil {
```

Comment:
why do we skip cleanup on ebpf but not iptables? I would think the logic should be consistent?

Reply:
This is to make sure traffic will still go through ztunnel if istio-cni is broken/stopped.   Without this, if istio-cni is stopped working, the existing traffic to ztunnel will be bypassed as well.

Reply:
There is also (3) CNI is stopped because its being replaced. We should NOT clean up here

Reply:
@howardjohn 👍  - we have no mechanism to determine (3) yet tho - that's what you mean by `latching` I guess.

Reply:
Yep exactly. Strawman proposal would be just to write to the host disk (we already do this for some other settings). but we may need ot think it through more

Reply:
Let me have an summary for _"stopping"_ case we discussed above:
1. CNI _crashes_ or terminates abnormally  --> everything may be indeterminate, new cni POD should restore the desired state
2. CNI is _graceful stopped_ due to _uninstall_   --> we'd like to do cleanup
3. CNI is _graceful stopped_ due to _upgrade/replacement_   --> we'd like to latch rules  to make sure no escaping during upgrade

So question comes to how to identify it's _uninstall_ or _upgrade/replacement_. Since there's no mechanism, I think we could move the _cleanup_ stuff  into `istioctl` or `helm`  `pre-delete` job and do nothing for the stop procedure of istio-cni process.

Reply:
>So question comes to how to identify it's uninstall or upgrade/replacement. Since there's no mechanism, I think we could move the cleanup stuff into istioctl or helm pre-delete or post-delete job and do nothing for the stop procedure of istio-cni process.

The problem is that the cleanup logic is tightly tied to the creation logic, and like the creation logic is complex/specific/requires a lot of the same support tooling.

Cleanup and teardown should live in the same spot. Separating them seems like a source of bad complexity. Rule creation cannot live "outside" of the CNI watch server, by definition. The problem is the fact that our CNI watchserver manages itself and is unaware of upgrades.

I kinda don't like Helm lifecycle hooks, as they tend to open up a Pandora's box of manually overriding resource creation order, manually garbage-collecting hook-created resources, etc etc. But if we are careful this is an option. The other problem with this is that as mentioned above we would essentially be duplicating much of the rule creation code in the `teardown` hook init/teardown container binary.

In much the same way that I am suspicious of Helm lifecycle hooks, I am suspicious of bespoke CRDs, but some of what Costin talked about https://github.com/istio/istio/issues/43642 WRT dropping the CNI plugin model  and just creating a CRD that manages rules and upgrades begins to make sense here.

Reply:
If we could find a good mechanism to identify it's an uninstall or upgrade/crash/terminating event with the daemonset itself,  we should cleanup everything at the same context created it.

Regarding to CRD, IMHO, I think the main idea of it is also to have the high dimension view of what exact task we need to do, it's more like a 'scavenger' operator who is the first and last one of all.

Reply:
My original comment is still not addressed. We only cleanup for ebpf but not iptables. Regardless of all other details, iptables and eBPF should follow the same cleanup

Reply:
Specifically AnnotateUnenrollPod